Remove all precedence from type system

tamasvajk opened the pull request:
This PR makes some fundamental changes to the type system. All precedence specifications are removed. The type system is made tighter, so `ref ref int` is not valid any longer, similarly to `int??`, which was already not allowed. Furthermore the parameter in `void M(ref ref int i){}` is also not parsing correctly any longer. Additionally, pointer types are removed from nullable base types, so `int*?` is not allowed any longer.

damieng commented:
Sorry, I cleaned up some of the unused conflicts earlier - I can revert that for now if that helps you get unblocked.

Do you think these changes might help us move forward with addressing some of the other open precedence bugs?

tamasvajk commented:
> Sorry, I cleaned up some of the unused conflicts earlier - I can revert that for now if that helps you get unblocked.

No problem. I already rebased.

> Do you think these changes might help us move forward with addressing some of the other open precedence bugs?

Yes, that's my hope. It will definitely help in introducing `scoped` types, and hopefully I can also remove the precedence on `modifier`, which might help with introducing `file` modifier.

tamasvajk commented:
Ohh, for the `file` modifier, I found an easier solution: https://github.com/tree-sitter/tree-sitter-c-sharp/pull/257

tamasvajk commented:
Looks like this change helps with the optional return type of lambdas: https://github.com/tamasvajk/tree-sitter-c-sharp/pull/2.

tamasvajk commented:
I've done some more tests on this by adding the below to `fetch-examples` and checking the number of errors. Both before and after these changes, the count is 177. So I don't think this PR introduces any (major) regression.
```
checkout examples/roslyn dotnet/roslyn ac92012c0e64c8114ecf0024c2bad2f63dcc7c6b
```

damieng commented:
I wonder if we should add Roslyn to the CI test suite.

tamasvajk commented:
> I wonder if we should add Roslyn to the CI test suite.

That would be great, but we would need some improvement to `parse-examples`, because if a known failure is fixed, we are not forced to remove the entry from `known-failures.txt`.
